packages/datadog-plugin-aws-sdk/src/services/kinesis.js

```diff
@@ -1,15 +1,64 @@
 'use strict'
-
+const log = require('../../../dd-trace/src/log')
 class Kinesis {
   generateTags (params, operation, response) {
-    const tags = {}
-
-    if (!params || !params.StreamName) return tags
+    if (!params || !params.StreamName) return {}
 
-    return Object.assign(tags, {
+    return {
       'resource.name': `${operation} ${params.StreamName}`,
       'aws.kinesis.stream_name': params.StreamName
-    })
+    }
+  }
+
+  // AWS-SDK will b64 kinesis payloads
+  // or will accept an already b64 encoded payload
+  // This method handles both
+  _tryParse (body) {
+    try {
+      return JSON.parse(body)
+    } catch (e) {
+      log.info('Not JSON string. Trying Base64 encoded JSON string')
+    }
+    try {
+      return JSON.parse(Buffer.from(body, 'base64').toString('ascii'), true)
+    } catch (e) {
+      return null
+    }
+  }
+
+  requestInject (span, request, tracer) {
+    const operation = request.operation
+    if (operation === 'putRecord' || operation === 'putRecords') {
+      if (!request.params) {
+        return
+      }
+
+      const traceData = {}
+      tracer.inject(span, 'text_map', traceData)
+      let injectPath
+      if (request.params.Records && request.params.Records.length > 0) {
+        injectPath = request.params.Records[0]
+      } else if (request.params.Data) {
+        injectPath = request.params
+      } else {
+        log.error('No valid payload passed, unable to pass trace context')
+        return
+      }
+      const byteSize = Buffer.byteLength(injectPath.Data, 'base64')
+      // Kinesis max payload size is 1MB
+      // So we must ensure adding DD context won't go over that (512b is an estimate)
+      if (byteSize + 512 >= 1000000) {
```

Comment:
The comment is definitely clearer now, but 1MB is not 1000000 bytes and also why do we need to estimate, can't we check the result after adding our metadata to make sure it still fits? It's a bit more wasteful but it's unlikely to happen often and it would be more accurate.

Reply:
Sure we could use `1048576` as the value here, seems worthwhile given our own context is small. And I'm happy to measure after, it'll be much more variable now given the new `x-datadog-tags` header